[docs] Requiring System Libraries with .systemLibrary targets (#6634)
apple/swift-package-manager: Documentation/Usage.md

## Requiring System Libraries

You can link against system libraries using the package manager. To do so, you'll
need to add a special `target` of type `.systemLibrary`, and a `module.modulemap`
for each system library you're using.

> Let's see an example of using [libgit2](https://libgit2.github.com) from an
> executable.

Reviewer: ```suggestion
Let's see an example of adding [libgit2](https://libgit2.github.com) as a
dependency to an executable target.
```

been installed by a system packager (eg. `apt`, `brew`, `yum`, etc.). The
following files from the libgit2 system-package are of interest:

    /opt/homebrew/lib/libgit2.dylib      # .so on Linux

>     /opt/homebrew/include/git2.h

Reviewer 2: I would probably leave this as `/usr/local` since not everyone is using homebrew, and change the note to say it may be in `/opt/homebrew` etc

Author: Cleaned this up in this and other spots.

**Note:** the system library may be located elsewhere on your system, such as
`/usr/local/`, or `/usr/`. The paths above are valid for Homebrew on macOS.

Now let's define `Clibgit` dependency in your `Package.swift` in the example project:

```swift
// swift-tools-version: 5.8
// The swift-tools-version declares the minimum version of Swift required to build this package.

import PackageDescription

let package = Package(
    name: "example",
    targets: [
        // Targets are the basic building blocks of a package, defining a module or a test suite.
        // Targets can depend on other targets in this package and products from dependencies.

>         .executableTarget(

Reviewer 2: it may be easier to understand if we do this in two steps, first introduce the "Clibgit" target, then show how to use it in the executable target. in other words, I suggest we split this example into two snippets, and build the knowledge in layers

Author: It did feel a little bit overwhelming when I tried to reproduce the steps, so I agree. Rewritten that in two steps, and cleaned up the `pkg-config` mention as well.

            name: "example",

            // example executable requires "Clibgit" target as it's dependency.
            // It's a systemLibrary target defined below.
            dependencies: ["Clibgit"],
            path: "Sources"
        ),

        // systemLibrary is a special type of build target that wraps a system library
        // in a target that other targets can require as their depencency.
        .systemLibrary(
            name: "Clibgit",
            pkgConfig: "libgit2",
            providers: [
                .brew(["libgit2"]),
                .apt(["libgit2-dev"])
            ]
        )
    ]
)

```

The `pkgConfig` parameter helps SwiftPM in figuring out the include and library
search paths for the system library. You can look up pkg-config for a package
by running the following:

    example$ pkg-config --cflags libgit2

>     -I/opt/homebrew/Cellar/libgit2/1.6.4/include

Reviewer 2: I am not sure what this means. can you elaborate what this is about?

Author: Moving this into the note about system paths, where it makes more sense. My point was that on most systems, for most libraries, running `pkg-config --cflags {library-name}` will get you the `inclide` and `lib` paths.

**Note:** If you don't want to use the `pkgConfig`
parameter you can pass the path of a directory containing the library using the
`-L` flag in commandline when building your app:

    example$ swift build -Xlinker -L/opt/homebrew/lib/

Next, we'll create a directory `Sources/Clibgit` in your `example` projcet, and
add a `module.modulemap` to it:

    module Clibgit [system] {

>       header "/opt/homebrew/include/git2.h"

Reviewer 2: ditto re. `/usr/local` with note that homebrew users this may be `/opt/homebrew/`

    /usr/local/lib/libgit2.dylib      # .so on Linux
    /usr/local/include/git2.h

**Note:** the system library may be located elsewhere on your system, such as

> `/usr/`, or `/opt/homebrew/` if you're using Homebrew on an Apple silicon Mac.

Reviewer: ```suggestion
`/usr/`, or `/opt/homebrew/` if you're using Homebrew on an Apple Silicon Mac.
```

Author: 🤦🏼 yikes, sorry about that.

If you're not sure where `Clibgit` is in your system, you can run this command
to look it up:

    example$ pkg-config --cflags libgit2
    -I/usr/local/libgit2/1.6.4/include

**First, let's define the `target` in the package description**:

```swift
// swift-tools-version: 5.8
// The swift-tools-version declares the minimum version of Swift required to build this package.

import PackageDescription

let package = Package(
    name: "example",
    targets: [
        // systemLibrary is a special type of build target that wraps a system library
        // in a target that other targets can require as their depencency.
        .systemLibrary(
            name: "Clibgit",
            pkgConfig: "libgit2",
            providers: [
                .brew(["libgit2"]),
                .apt(["libgit2-dev"])
            ]
        )
    ]
)

```

**Note:** If you don't want to use the `pkgConfig` parameter you can pass the
path of a directory containing the library using the `-L` flag in command line
when building your app instead:

    example$ swift build -Xlinker -L/usr/local/lib/

> Next, we'll create a directory `Sources/Clibgit` in your `example` projcet, and

Reviewer: Nit: typo, I'd also avoid the use of "we" here to stay consistent with using "you" in other paragraphs.
```suggestion
Next, create a directory `Sources/Clibgit` in your `example` project, and
```

add a `module.modulemap` to it:

    module Clibgit [system] {

>       header "/usr/local/include/git2.h"

Reviewer 2: there is a somewhat better way to do this, namely to create you own header file, which then include the required one with `<header>` so that no path hardcoding is required.  eg: https://github.com/apple/swift-package-manager/tree/main/Sources/SPMSQLite3

Author: Sorry, long flight — back in PST now. 

Ah, that's a TIL for me. So it looks like:
- You can, and probably should, add a header file in Sources/{LibName}/{LibName.h} that just includes {LibName.h} and/or other required headers. 
- The `#include` will resolve with the compiler environment, so it should pick up the headers correctly, and thus you don't have to write the absolute header path in the `module.modulemap`. 

Right?

Reviewer 2: yup afaik. @neonichu to keep me honest here

    /usr/local/lib/libgit2.dylib      # .so on Linux
    /usr/local/include/git2.h

> **Note:** the system library may be located elsewhere on your system, such as

Reviewer 3: I think that we should clear up some of the text here.  Windows is a support platform, and these paths are not relevant there.  Additionally, `pkg-config` is not a commonly available utility on Windows.

Author: Added an entry on Windows paths.

`/usr/`, or `/opt/homebrew/` if you're using Homebrew on an Apple Silicon Mac.
If you're not sure where `Clibgit` is in your system, you can run this command
to look it up:

    example$ pkg-config --cflags libgit2
    -I/usr/local/libgit2/1.6.4/include

**First, let's define the `target` in the package description**:

```swift
// swift-tools-version: 5.8
// The swift-tools-version declares the minimum version of Swift required to build this package.

import PackageDescription

let package = Package(
    name: "example",
    targets: [
        // systemLibrary is a special type of build target that wraps a system library
        // in a target that other targets can require as their depencency.
        .systemLibrary(
            name: "Clibgit",
            pkgConfig: "libgit2",
            providers: [
                .brew(["libgit2"]),

>                 .apt(["libgit2-dev"])

Reviewer 3: Perhaps add a `.nuget` entry as well?

Author: Hmmm. Should we suggest [`LibGit2Sharp`](https://www.nuget.org/packages/LibGit2Sharp)? Swift doesn't need the .Net wrapper to build it's package, but there doesn't seem to be another minimal nuget package for it.

            ]
        )
    ]
)

```

> **Note:** If you don't want to use the `pkgConfig` parameter you can pass the

Reviewer 3: `pkgConfig` should be elided for Windows only packages as `pkg-config` is not expected to be available.

Author: Added that to the note.

path of a directory containing the library using the `-L` flag in command line
when building your app instead:

    example$ swift build -Xlinker -L/usr/local/lib/

Next, create a directory `Sources/Clibgit` in your `example` project, and
add a `module.modulemap` and the header file to it:

    module Clibgit [system] {
      header "git2.h"
      link "git2"
      export *
    }

The header file should look like this:

```c
# git2.h

> #pragma once

Reviewer 3: Should we recommend include guards instead?

Author: I honestly don't have enough experience with C, so I don't know if using `#pragma once` is perfectly fine, or a bad taste. My understanding is that using guards is guaranteed to work anywhere, and using `#pragma once` is shorter, but might not work in certain environments?

print(jpegData)
```

Install JPEG library using a system packager, e.g, `$ brew install jpeg`.
`jpeg` is a keg-only formula, meaning it won't be linked to `/opt/homebrew/lib`,
and you'll have to link it manually at build time.

Just like in the previous example, `mkdir Sources/CJPEG`, and add the
following `module.modulemap`:

    module CJPEG [system] {
        header "shim.h"

>         header "/opt/homebrew/opt/jpeg/include/jpeglib.h"

Reviewer 2: ditto re. `/usr/local` with note that homebrew users this may be `/opt/homebrew/`

    git_repository_init_options(version: 0, flags: 0, mode: 0, workdir_path: nil, description: nil, template_path: nil, initial_head: nil, origin_url: nil)
    example$

> ### Requiring a System Library Without pkg-config

Reviewer: ```suggestion
### Requiring a System Library Without `pkg-config`
```

print(jpegData)
```

> Install JPEG library using a system packager, e.g, `$ brew install jpeg`.

Reviewer: ```suggestion
Install JPEG library using a Homebrew package manager with `brew install jpeg`.
```

> Install JPEG library, on macOS you can use Homebrew package manager: `brew install jpeg`.

Reviewer: ```suggestion
Install the JPEG library, on macOS you can use Homebrew package manager: `brew install jpeg`.
```

`jpeg` is a keg-only formula, meaning it won't be linked to `/usr/local/lib`,
and you'll have to link it manually at build time.

> Just like in the previous example, `mkdir Sources/CJPEG`, and add the

Reviewer: ```suggestion
Just like in the previous example, run `mkdir Sources/CJPEG` and add the
```

the required line `#include <stdio.h>`. Alternatively, you can add `#include <stdio.h>`
to the top of jpeglib.h to avoid creating the `shim.h` file.

Now to use the CJPEG package we must declare our dependency in our example
app’s `Package.swift`:

```swift
// swift-tools-version: 5.8

> // The swift-tools-version declares the minimum version of Swift required to build this package.

Reviewer 2: this line is redundant

import PackageDescription

let package = Package(
    name: "example",
    targets: [
        // Targets are the basic building blocks of a package, defining a module or a test suite.

>         // Targets can depend on other targets in this package and products from dependencies.

Reviewer 2: the comment lines are redundant

        .executableTarget(
            name: "example",
            dependencies: ["CJPEG"],
            path: "Sources"
            ),

>         .systemLibrary(

Reviewer 2: ditto re. splitting the example into two steps

Author: After splitting the first one, I'm not so sure if we should split this one as well. It would add another blob of code, but since the second example builds up on the first one, I think the concept should be clear even without splitting. 

But, happy to split if you'd prefer it in two steps here as well!

            name: "CJPEG",
            providers: [
                .brew(["jpeg"])
            ])
    ]
)
```

Now if we type `swift build` in our example app directory we will create an
executable:

>     example$ swift build -Xlinker -L/opt/homebrew/opt/jpeg/lib

Reviewer 2: ditto re. `/usr/local` with note that homebrew users this may be `/opt/homebrew/`

print(jpegData)
```

Install the JPEG library, on macOS you can use Homebrew package manager: `brew install jpeg`.
`jpeg` is a keg-only formula, meaning it won't be linked to `/usr/local/lib`,
and you'll have to link it manually at build time.

Just like in the previous example, run `mkdir Sources/CJPEG` and add the
following `module.modulemap`:

    module CJPEG [system] {
        header "shim.h"

>         header "/usr/local/opt/jpeg/include/jpeglib.h"

Reviewer 3: Should we recommend something with a VFS example?  That would make the majority of the code more cross-platform friendly.

Author: For this and all your other comments, thank you! I'll catch up and clean things up a bit later today (PST) and ping you when it's all updated.

the package manager will pick the manifest with the most compatible tools
version. For example, if there are three manifests:

> `Package.swift` (tools version 3.0)  

Author: These three lines are non-intentional, my formatter got trigger-happy. I think that's okay?
